### DISCUSSION TITLE
ReferenceError: crypto is not defined

### DISCUSSION THREAD

**Discussion by rbrt-stn** [darkreader/darkreader#7349] [Q&A]
Hello, 

I am using DarkReader in my own React application. 
Everything works and looks great, except for the fact that some of my already-written tests are failing with the error: 
`ReferenceError: crypto is not defined`

I tracked the problem and it seems like the issue is in the function generateUID() from darkreader.js, which gets called at the startup of my tests, since I use the darkreader library in a Context. 

To solve the issue in my project, I tried to manually 
`npm install crypto`
but this didn't work.

Has anyone else faced this issue? 

Edit: I think it's related to this: https://stackoverflow.com/questions/52732358/typescript-libraries-are-not-available-when-running-mocha-tests

**Comment by Gusted** [comment]
Hi @rbrt-stn!

I think you don't have the correct test Environment, `crypto` is a web-api. And `mocha` is by-default just using the node context.
You've to enable a `jsdom` environment, e.g. https://www.npmjs.com/package/mocha-jsdom

Regards,
Gusted

**Reply by rbrt-stn** [reply]
Thank you for your answer! Unfortunately your solution doesn't work. 

However, I solved the issue by globally mocking the crypto in setupTests.js
